refactor(backfill): Backfill Plugin Major Refactor And Improvements

AlfredoG87 opened the pull request:
## Summary

  Refactors the `BackfillPlugin` from a monolithic ~650-line class into a modular architecture of focused, testable components.

  **Key change:** Introduces a **dual-scheduler design** that separates historical and live-tail backfill processing—ensuring recent gaps are never blocked by long-running historical catch-up.

  ---

  ## Architecture Overview

  **`BackfillPlugin` now acts as an orchestrator** coordinating three stages:

  ### 1. **Detect** — `GapDetector`
  - Scans storage for missing block ranges
  - Classifies gaps as **HISTORICAL** (older) or **LIVE_TAIL** (recent/near-head)
  - Simplifies Greedy Backfill by defining an upper bound based on true/false -> peerMax/latestStoredBlock

  ### 2. **Schedule** — `BackfillTaskScheduler`
  - Routes gaps into dedicated bounded queues by type
  - Historical and live-tail backfills run independently

  ### 3. **Execute** — `BackfillRunner`
  - Selects optimal peer via `PriorityHealthBasedStrategy`
  - Fetches blocks in configurable batches
  - **Awaits persistence confirmation before fetching more** (natural backpressure)

  ---

  ## New Components

  | Class | Responsibility |
  |-------|----------------|
  | `BackfillRunner` | Core execution logic |
  | `BackfillTaskScheduler` | Bounded queue task scheduling |
  | `BackfillPersistenceAwaiter` | Persistence-based backpressure |
  | `GapDetector` | Gap detection and classification |
  | `PriorityHealthBasedStrategy` | Health-aware node selection |
  | `NodeSelectionStrategy` | Node selection abstraction |

  ---

  ## Configuration

  | Property | Default | Description |
  |----------|---------|-------------|
  | `backfill.historicalQueueCapacity` | 20 | Max pending gaps for historical scheduler |
  | `backfill.liveTailQueueCapacity` | 10 | Max pending gaps for live-tail scheduler |
  | `backfill.healthPenaltyPerFailure` | 1000.0 | Health score penalty per node failure |
  | `backfill.maxBackoffMs` | 300000 | Maximum backoff duration (5 min) |

  ---

  ## Other Changes

  - Exposed WebClient HTTP/2 tuning parameters with high-throughput defaults
  - Extended `BlockNodeSource` proto with optional `NodeId` and `Name` fields (non-breaking)
  - Updated design and configuration documentation
  - Reduced annoyingly large logs in the MessagingFacility Impl
---

## Review guide (recommended order)

**If you want the fastest mental model, I recommend reviewing in this order:**

1. GapDetector – gap identification + HISTORICAL vs LIVE_TAIL classification
2. BackfillTaskScheduler – dual bounded queues + worker lifecycle
3. BackfillRunner – node selection → fetch → dispatch → await persistence loop
4. BackfillPersistenceAwaiter – how backpressure is enforced via notifications
5. Config + docs – new properties, defaults, and documentation updates

  ---

## PR Stats
  | Category    | Lines  | % of PR |
  |-------------|--------|---------|
  | Tests       | ~2,459 | 70%     |
  | Main Source | ~1,282 | 36%     |
  | Docs        | ~261   | 7%      |

  - Tests (70%): 102 test methods covering critical backfill infrastructure. 
  - Main Source (36%): Significant architectural refactor - BackfillPlugin split into 6 focused components (BackfillRunner, GapDetector, TaskScheduler, etc.)
  - Docs (7%): Updated design documentation

---

## Related Issues
Fixes #1977  
Fixes #1550  
Fixes #1502  
Fixes #1778

ata-nas commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/client/proto/block_node_source.proto:
```diff
@@ -11,4 +11,51 @@ message BackfillSourceConfig {
+  bool ping_enabled = 10;
+
+  // HTTP/2 ping timeout in milliseconds (default: 10000)
+  uint32 ping_timeout = 11;
```
Just noting, the receiving side should also have some sane boundaries as to what it can accept as values, otherwise, it should probably rollback to defaults.

AlfredoG87 replied:
by receiving side you mean also the BN (Server), we do have these boundaries as part of the config min max constrains.

AlfredoG87 replied:
I can't confirm, since I am not entirely sure what you mean here.

This is where we set the other side of the communication the webServer: 
https://github.com/hiero-ledger/hiero-block-node/blob/main/block-node/app/src/main/java/org/hiero/block/node/app/BlockNodeApp.java#L183-L218

However what we are exposing here is a way to fine tune each webClient for each node according to their own network.

For the webServer side of the connection we have them as a regular system config, thus with its own `max`,  `min` and default values:
https://github.com/hiero-ledger/hiero-block-node/blob/main/block-node/app-config/src/main/java/org/hiero/block/node/app/config/ServerConfig.java
https://github.com/hiero-ledger/hiero-block-node/blob/main/block-node/app-config/src/main/java/org/hiero/block/node/app/config/WebServerHttp2Config.java

ata-nas replied:
@jsync-swirlds ah, thanks for the note, I had that confused.

ata-nas commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/client/BlockNodeClient.java:
```diff
@@ -32,55 +39,95 @@ private static record Options(Optional<String> authority, String contentType)
+    }
 
-        initializeClient();
+    private int getOrDefault(
```
As mentioned above, it is probably a good idea to have a range defined here for each config, which is sane and if the provided value is outside of that range, the default value could be used.

Moreover, I think in some places validations are also due, like in values where powers of 2 are needed etc.

This can again be added as a simple predicate which holds the validation of the value, if failed, then default is returned.

ata-nas commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillFetcher.java:
```diff
@@ -343,7 +323,89 @@ public List<BlockUnparsed> fetchBlocksFromNode(BackfillSourceConfig nodeConfig,
+        merged.add(current);
+
+        return merged;
+    }
```
This is probably best extracted to a `RangeSupport` class. In the implementation of the ranged sets there are methods that do very efficient merging. Something to consider.

AlfredoG87 replied:
I've extracted this to LongRange class itself, good catch! 👍 

✅ Done!

ata-nas replied:
Not sure how I feel about that. I think we should probably keep the LongRange class as simple as possible. Merging ranges does not feel like something the LongRange class should be responsible for.

AlfredoG87 replied:
I considered using ConcurrentLongRangeSet but chose the static utility for these reasons:

  1. No concurrency needed here - getAvailabilityForRange() runs single-threaded, so ConcurrentLongRangeSet's thread-safety overhead isn't beneficial
  2. One-shot operation - We're not maintaining a mutable collection; we collect ranges, merge once, and return an immutable result
  3. LongRange.mergeContiguousRanges() is simpler - It's a pure function that takes a list and returns a merged list, no state management
  4. The merge logic is now in LongRange (moved from BackfillFetcher), so it's reusable across the codebase without pulling in ConcurrentLongRangeSet

ata-nas commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java:
```diff
@@ -223,322 +167,225 @@ public void start() {
+        // Deduplicate live-tail gaps using high-water mark
+        GapDetector.Gap effectiveGap = gap;
+        if (gap.type() == GapDetector.Type.LIVE_TAIL) {
+            long hwm = liveTailHighWaterMark.get();
```
The abbreviation is really confusing. Could we spell high-water mark?

ata-nas commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java:
```diff
@@ -223,322 +167,225 @@ public void start() {
-            LOGGER.log(TRACE, "Fetching blocks from start={0} to end={1}", gap.start(), gap.end());
-            backfillGap(gap, BackfillType.AUTONOMOUS);
+        BackfillTaskScheduler scheduler =
+                (effectiveGap.type() == GapDetector.Type.HISTORICAL) ? historicalScheduler : liveTailScheduler;
```
This could be a method.

AlfredoG87 replied:
Extracted to its own method. and added comments.

✅ Done!

ata-nas commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillRunner.java:
```diff
@@ -0,0 +1,174 @@
+     * @throws InterruptedException if the thread is interrupted
+     */
+    void run(@NonNull GapDetector.Gap gap) throws ParseException, InterruptedException {
+        backfillGap(gap.range(), gap.type());
```
Why not have the logic of the backfillGap method directly in here? That method seems redundant.

AlfredoG87 replied:
To provide a clean separation between the runners run and it's internal logic, also detectGaps is a vestigial part of the BackfillPlugin that was extracted to the BackfillRunner, so it makes sense to keep it separated even if it is a little redundant.

ata-nas replied:
Still, personally I do not see any reason to do so, with the code we have. This is nothing blocking, just a suggestive nit.

ata-nas commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/GapDetector.java:
```diff
@@ -0,0 +1,95 @@
+     * and overlapping gaps are split.
+     */
+    @NonNull
+    List<Gap> findTypedGaps(
```
What does TypedGap mean?

AlfredoG87 replied:
Nowadays every gap has a type, but originally this was not the case so it was meant to be that the Gap has a Type.

It could just say findGaps but there is already a private method that finds the raw gaps as LongRange that has that name.

This one consumes findGaps and classifies them into its respective Type. (either Historical or Live-Tail) hence TypedGaps method name makes sense.

ata-nas commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/PriorityHealthBasedStrategy.java:
```diff
@@ -0,0 +1,164 @@
+            long startBlock, long gapEnd, @NonNull Map<BackfillSourceConfig, List<LongRange>> availability) {
+        long earliestAvailableStart = Long.MAX_VALUE;
+        for (Map.Entry<BackfillSourceConfig, List<LongRange>> entry : availability.entrySet()) {
+            for (LongRange availableRange : entry.getValue()) {
```
You could just do availability.values() instead of this nested for loop.

ata-nas reviewed (commented):
First pass on this one, will need at least another one. Have not yet looked at tests as well.

I leave general comments for cleanup and questions.

I feel, however, that we are really missing out on not using the ranged sets and instead we are using a list of ranges and manually manipulating them. I do see usages of merging ranges, this is something the ranged sets will do automatically in a more performant fashion. I guess I just do not see why we need to use a list and complicating some of the things we do, instead of just using the ranged sets. Since we re doing such a big rework, now is the time to make decisions. Something worth thinking about.

AlfredoG87 commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/client/BlockNodeClient.java:
```diff
@@ -32,55 +39,95 @@ private static record Options(Optional<String> authority, String contentType)
+
+    private Http2ClientProtocolConfig buildHttp2Config(@Nullable GrpcWebClientTuning tuning) {
+        // Skip HTTP/1.1 upgrade negotiation for better performance
+        final boolean priorKnowledge = tuning == null || tuning.priorKnowledge();
```
For boolean values we don't need to use the getOrDefault, I believe this way is clearer and leaner.

ata-nas replied:
I am OK with that, I was just wondering if we can remove all the cumbersome statements into a clean method. Not blocking, just a nit.

AlfredoG87 commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillConfiguration.java:
```diff
@@ -39,4 +39,10 @@ public record BackfillConfiguration(
+        @Loggable @ConfigProperty(defaultValue = "10") @Min(1) @Max(100) int liveTailQueueCapacity,
+        // Health scoring constants
+        @Loggable @ConfigProperty(defaultValue = "1000.0") double healthPenaltyPerFailure,
+        @Loggable @ConfigProperty(defaultValue = "300000") @Min(30000) long maxBackoffMs) {}
```
I've added this nit on all configs where it was possible.

✅  Done!

ata-nas replied:
Ahh...

AlfredoG87 commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillFetcher.java:
```diff
@@ -343,7 +323,89 @@ public List<BlockUnparsed> fetchBlocksFromNode(BackfillSourceConfig nodeConfig,
+        double failurePenalty = h.failures * healthPenaltyPerFailure;
+        double latencyPenaltyMs = h.successes > 0 ? (h.totalLatencyNanos / (double) h.successes) / 1_000_000.0 : 0;
+        return failurePenalty + latencyPenaltyMs;
+    }
```
I've added how it works on the backfill-design doc.

✅ Done!

AlfredoG87 commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPersistenceAwaiter.java:
```diff
@@ -0,0 +1,151 @@
+
+        LOGGER.log(TRACE, "Waiting for block [%s] persistence (timeout=[%s]ms)".formatted(blockNumber, timeoutMs));
+        boolean completed = latch.await(timeoutMs, TimeUnit.MILLISECONDS);
+        pendingBlocks.remove(blockNumber);
```
I've refactored awaitPersistence to handle InterruptedException and finally always removing the pending block.

✅ Done!

AlfredoG87 commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPersistenceAwaiter.java:
```diff
@@ -0,0 +1,151 @@
+     *
+     * @param blockNumber the block number to wait for
+     * @param timeoutMs maximum time to wait in milliseconds
+     * @return true if persistence was confirmed, false if timed out
```
You are very right here, even if it never existed it  would return true, which is desired, but the comment is not clear about it.

I've improved the wording, thanks! 👍 

✅ Done!

AlfredoG87 commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java:
```diff
@@ -223,322 +167,225 @@ public void start() {
+        List<LongRange> blockRanges = context.historicalBlockProvider()
+                .availableBlocks()
+                .streamRanges()
+                .toList();
```
I am trying to keep PR contained to the Backfill plugin only to prevent conflicts, this is a good idea but rather will leave it for the next time 😅

AlfredoG87 commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillRunner.java:
```diff
@@ -0,0 +1,174 @@
+    private void backfillGap(LongRange gap, GapDetector.Type gapType) throws InterruptedException, ParseException {
+        logger.log(TRACE, "Starting backfillGap type=[%s] range=[%s]".formatted(gapType, gap));
+        Map<BackfillSourceConfig, List<LongRange>> availability = planAvailabilityForGap(fetcher, gap);
+        long currentBlock = gap.start();
```
We discussed this, yes, but we decided to not go that route. we will always backfill ascending.

AlfredoG87 commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/PriorityHealthBasedStrategy.java:
```diff
@@ -0,0 +1,164 @@
+            @NonNull Map<BackfillSourceConfig, List<LongRange>> availability) {
+        List<NodeSelection> candidates = new ArrayList<>();
+        for (Map.Entry<BackfillSourceConfig, List<LongRange>> entry : availability.entrySet()) {
+            for (LongRange availableRange : entry.getValue()) {
```
Not quite the same as above, but left a comment.

✅ Done!

AlfredoG87 commented:
@ata-nas Thank you for reviewing my PR, really appreciate and value all your input and time effort made into reviewing it 🙏 

I've addressed all your comments mostly in a positive outcome.

And for your general notes suggestion:
> I feel, however, that we are really missing out on not using the ranged sets and instead we are using a list of ranges and manually manipulating them. I do see usages of merging ranges, this is something the ranged sets will do automatically in a more performant fashion. I guess I just do not see why we need to use a list and complicating some of the things we do, instead of just using the ranged sets. Since we re doing such a big rework, now is the time to make decisions. Something worth thinking about.

I've ponder this up and decided that the benefits do not outweigh the effort and added complexity. The use case we need is simpler as opposed to the BlockRangeSet impls and prefer to keep it simple as a native List.

ata-nas commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java:
```diff
@@ -176,38 +126,24 @@ public void init(BlockNodeContext context, ServiceBuilder serviceBuilder) {
                     TRACE,
-                    "Block node sources path file does not exist: {0}, backfill will not run",
-                    backfillConfiguration.blockNodeSourcesPath());
+                    "Block node sources path file does not exist: [%s], backfill will not run"
```
Nit: checking for regular file will return false if the file does not exist, is not a regular file, or it cannot be determined if the file is a regular file or not. The log could reflect that for accuracy, otherwise we are going to assume the file is not there based on the log.

ata-nas commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillFetcher.java:
```diff
@@ -58,40 +59,75 @@ public class BackfillGrpcClient {
 
         for (BackfillSourceConfig node : blockNodeSource.nodes()) {
-            LOGGER.log(INFO, "Address: {0}, Port: {1}, Priority: {2}", node.address(), node.port(), node.priority());
+            GrpcWebClientTuning tuning = node.grpcWebclientTuning();
```
If we are sure about non-null values, then we should be OK, still, I would personally check the input every time before we propagate the objects down the line.

AlfredoG87 commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/client/BlockNodeClient.java:
```diff
@@ -32,55 +39,95 @@ private static record Options(Optional<String> authority, String contentType)
+            @Nullable GrpcWebClientTuning tuning, ToIntFunction<GrpcWebClientTuning> getter, int defaultValue) {
+        if (tuning == null) return defaultValue;
+        int value = getter.applyAsInt(tuning);
+        return value > 0 ? value : defaultValue;
```
0 does not make sense for any of these values, and is PBJ default for numbers.
So is safe to use it as non-set.

AlfredoG87 commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java:
```diff
@@ -223,322 +159,240 @@ public void start() {
+                    queueCapacity,
+                    fetcher,
+                    persistenceAwaiter);
+        } catch (Exception e) {
```
I've removed all instances of Exception 
✅ Done!

jsync-swirlds reviewed (commented):
Still reviewing, but below are some thoughts on the updated design document.

AlfredoG87 commented on docs/design/backfill-plugin.md:
```diff
@@ -4,257 +4,308 @@
-            Note right of BackfillPlugin: nothing to backfill
+            MF->>Persist: Verify & persist
+            Persist-->>BP: PersistedNotification
         end
```
It says Verify and Persist, for simplification purposes to make it lighter I think is not necessary. but I could be wrong.

AlfredoG87 commented on docs/design/backfill-plugin.md:
```diff
@@ -4,257 +4,308 @@
-  <dt>BlockSource</dt>
-  <dd>A new Enum that will be added to existing notification types: `VerificationNotification` and `PersistedNotification` to indicate the original source where the block is coming from, currently it will only have two values: `Publisher`, `Backfill`</dd>
-
-</dl>
```
I've re-introduced the terms, but improved the list by sorting them and adding some more.

✅ Done!

AlfredoG87 commented on docs/design/backfill-plugin.md:
```diff
@@ -4,257 +4,308 @@
+      "address": "peer2.example.com",
+      "port": 8080,
+      "priority": 2,
+      "node_id": "peer-2",
```
✅ Done!

AlfredoG87 commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java:
```diff
@@ -223,322 +159,240 @@ public void start() {
+            historicalScheduler.close();
+        }
+        if (liveTailScheduler != null) {
+            liveTailScheduler.close();
```
I've added the try-catch.

✅ Done!

mustafauzunn commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillConfiguration.java:
Why we introduce all the new lines in this file?
It is harder to read and also breaks the common structure with other config files. Also if you need to read the javadoc you must scroll with the additional lines.

AlfredoG87 replied:
Yes I agree, but it was work of spotlessApply, the new Hiero conventions plugin forces that format, so for now we just have to live with it :(

jjohannes replied:
We use the https://github.com/palantir/palantir-java-format formatter. I am not sure of the history, but it was used in consensus node initially I believe. Unfortunately, how it behaves exactly in which cases is not documented. (Which could be a reason to rethink the choice of formatter at some point.)

It looks like annotations with arguments – e.g. `@ConfigProperty(defaultValue = "10")` – force a line break after the annotation. And then the fact that an argument declaration now consists of two lines, forces the additional empty line.  If there would only be annotation without arguments – e.g. `@Loggable` - the original formatting would be accepted.

Unfortunately, there is no way to reconfigure anything. It's a non-configurable formatter.

I agree that `// spotless:off` and `// spotless:on` is a good solution here.

AlfredoG87 replied:
Disabled spotless on BackfillConfiguration due to known issue with formatter with this file.

✅ Done!

ata-nas commented on block-node/backfill/src/test/java/org/hiero/block/node/backfill/BackfillFetcherTest.java:
```diff
@@ -0,0 +1,510 @@
+                mock(Counter.class), // backfillRetries
+                mock(LongGauge.class), // backfillStatus
+                mock(LongGauge.class), // backfillPendingBlocksGauge
+                mock(LongGauge.class)); // backfillInFlightGauge
```
We should not use mocks here at all. We already have the test helper which creates metrics which we can actually call to assert on. Please refer to: `TestUtils#createMetrics` and it's sample usages.

ata-nas commented on block-node/backfill/src/test/java/org/hiero/block/node/backfill/BackfillFetcherTest.java:
```diff
@@ -0,0 +1,510 @@
+            final BackfillFetcher client = newClient(lowPriority, highPriority);
+
+            // Selects earliest start even if higher priority
+            var selection = client.selectNextChunk(
```
Non-blocking nit: it would be much better, in my opinion, to never use var, considered everywhere. Optional, however.

AlfredoG87 replied:
Yes, I have made the changes too.

✅ Done!

ata-nas commented on block-node/backfill/src/test/java/org/hiero/block/node/backfill/BackfillFetcherTest.java:
```diff
@@ -0,0 +1,510 @@
+    class SelectNextChunkTests {
+
+        @Test
+        @DisplayName("selects earliest start, breaks ties by priority, returns empty when no match")
```
Nit: judging by the display name, these are 3 separate tests.

ata-nas commented on block-node/backfill/src/test/java/org/hiero/block/node/backfill/BackfillPersistenceAwaiterTest.java:
```diff
@@ -0,0 +1,412 @@
+            ExecutorService executor = Executors.newSingleThreadExecutor();
+            executor.submit(() -> {
+                try {
+                    Thread.sleep(50);
```
Unsure if 50ms is enough of a sleep. Maybe it finishes before we even start awaiting? Could be wrong though.

ata-nas commented on block-node/backfill/src/test/java/org/hiero/block/node/backfill/BackfillPersistenceAwaiterTest.java:
```diff
@@ -0,0 +1,412 @@
+            // then
+            assertEquals(3, count);
+        }
+    }
```
If we really consider `getPendingCount` to be a "test-only" method, should we test it? It feels like we would be better with a different approach here. Not sure.

ata-nas commented on block-node/backfill/src/test/java/org/hiero/block/node/backfill/BackfillPluginHelperTest.java:
```diff
@@ -21,29 +35,120 @@ private BackfillSourceConfig node(String host, int port) {
+
+            assertEquals(1, configTypes.size(), "Should return exactly one config type");
+            assertTrue(configTypes.contains(BackfillConfiguration.class), "Should contain BackfillConfiguration.class");
+        }
```
We generally accepted that we will not test configurations. It _should_ be enough to simply start the plugin in order to assert that the configuration is there, as it will be used, hopefully checked for non-null if simply stored as a field during init.

ata-nas commented on block-node/backfill/src/test/java/org/hiero/block/node/backfill/GapDetectorTest.java:
```diff
@@ -0,0 +1,112 @@
+        // Skips gaps where start > endCap
+        gaps = detector.findTypedGaps(List.of(new LongRange(0, 5), new LongRange(100, 200)), 0, 10, 3);
+        assertTrue(gaps.isEmpty());
+    }
```
Here, please note that as of now, the LongRange class does NOT support Long.MAX_VALUE as an end to the range! The reason being is the correct working of the size method thereof.

ata-nas reviewed (commented):
My second review of the PR.
This is only reviewing tests! (Pt.1, still have a couple of test classes to go through).
This took a very long time to review thoroughly.

As a general comment, I think we will really benefit of not using any mocks, but develop simple test fixtures. Moreover, for things such as metrics, we already have an approach to create them for tests w/o any mocks.

As mentioned in my previous review, this major rework is the best time to make the best decisions.

AlfredoG87 commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillFetcher.java:
```diff
@@ -0,0 +1,358 @@
+                        INFO,
+                        "Failed to get status from node [%s:%d]: %s"
+                                .formatted(node.address(), node.port(), e.getMessage()),
+                        e);
```
I've refactor all logging within the Backfill Plugin to keep the standard defined, of using the built-in string interpolation within the logger for performance reasons.

✅ Done!

AlfredoG87 commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillFetcher.java:
```diff
@@ -0,0 +1,358 @@
+        LOGGER.log(
+                TRACE,
+                "Determined available range from peer blocks nodes start=[%s] to end=[%s]"
+                        .formatted(startBlock, latestPeerBlock));
```
I've refactor all logging within the Backfill Plugin to keep the standard defined, of using the built-in string interpolation within the logger for performance reasons.

✅ Done!

AlfredoG87 commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPersistenceAwaiter.java:
I've refactor all logging within the Backfill Plugin to keep the standard defined, of using the built-in string interpolation within the logger for performance reasons.

✅ Done!

AlfredoG87 commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/client/proto/block_node_source.proto:
```diff
@@ -11,4 +11,69 @@ message BackfillSourceConfig {
+//   - HTTP/2 frame size: 16,384 bytes (16KB) to 16,777,215 bytes (~16MB) per RFC 7540
+//   - HTTP/2 window size: 1 to 2,147,483,647 (2^31-1) per RFC 7540
+//   - Buffer sizes: 1,024 bytes (1KB) to 67,108,864 bytes (64MB)
+//   - Header list size: 1,024 bytes (1KB) to 1,048,576 bytes (1MB)
```
Good to know this! 👍 

✅ Done!

AlfredoG87 commented on block-node/backfill/src/test/java/org/hiero/block/node/backfill/BackfillFetcherTest.java:
```diff
@@ -0,0 +1,510 @@
+    class GetAvailabilityForRangeTests {
+
+        @Test
+        @DisplayName("returns intersection for reachable nodes, empty when no overlap or unreachable")
```
Same as above, splited into multiple tests

✅ Done!

AlfredoG87 commented on block-node/backfill/src/test/java/org/hiero/block/node/backfill/BackfillFetcherTest.java:
```diff
@@ -0,0 +1,510 @@
+            var mismatchClient = mockClientReturning(List.of(createTestBlock(0L)));
+            fetcher = createFetcherWithClient(nodeConfig, 1, metrics, mismatchClient);
+            assertTrue(
+                    fetcher.fetchBlocksFromNode(nodeConfig, new LongRange(0, 1)).isEmpty());
```
Empty is intentional - when we request a range and get fewer blocks, we treat it as data inconsistency and fail (with markFailure). Returning partial results could create more gaps. Caller retries after backoff and source BN can be potentially backfilling itself from its own source, this means that it has gaps, we don't know how large or small until we have more details and implement the serverStatusDetails

AlfredoG87 commented on block-node/backfill/src/test/java/org/hiero/block/node/backfill/BackfillFetcherTest.java:
```diff
@@ -0,0 +1,510 @@
+
+            // Already ahead of peers
+            range = fetcher.getNewAvailableRange(100L);
+            assertNull(range);
```
Both return null because caller just needs "is there new work?" The reason (unreachable vs caught up) is logged but doesn't change action - both mean "nothing to do now."

ata-nas replied:
Got it. This part: `From the caller's perspective, the outcome is identical: this node cannot serve the requested range right now.` is key.

jsync-swirlds commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillFetcher.java:
```diff
@@ -0,0 +1,357 @@
+                                        maxRetries,
+                                        e.getMessage(),
+                                        e.getCause());
+                LOGGER.log(INFO, failedToFetchBlocksMsg, e);
```
Because you're trying to _not_ log stack trace, you may want to prefer the following:
```suggestion
                final String format = "Failed to fetch blocks [{0}->{1}] from node [{2}] (attempt {3}/{4}) due to {5}{6}.";
                final String cause = e.getCause() != null ? ", " + e.getCause() : "";
                LOGGER.log(INFO, message, blockRange.start(), blockRange.end(), nodeConfig.address(), attempt, maxRetries, e, cause);
```
This is a bit clearer, avoids formatting if not necessary, and avoids logging any stack traces.

For even more succinct logging, you might consider making this log statement `DEBUG` level.

jsync-swirlds commented on block-node/backfill/src/main/proto/block_node_source.proto:
```diff
@@ -0,0 +1,115 @@
+   * Valid range: 100ms to 300,000ms
+   */
+  uint32 ping_timeout = 11;
+}
```
Missing newline at EOF.

AlfredoG87 commented on block-node/backfill/src/test/java/org/hiero/block/node/backfill/BackfillPersistenceAwaiterTest.java:
```diff
@@ -0,0 +1,412 @@
+            subject.trackBlock(blockNumber);
+
+            // then
+            assertEquals(1, subject.getPendingCount());
```
I guess what they are asserting is somewhat redundant anyways, I've simplified the tests that involved getPendingCount that was indeed a test-only method (smell)

✅ Done!

AlfredoG87 commented on block-node/backfill/src/test/java/org/hiero/block/node/backfill/PriorityHealthBasedStrategyTest.java:
```diff
@@ -0,0 +1,282 @@
+    class PrioritySelectionTests {
+
+        @Test
+        @DisplayName("should select node with lowest priority number (highest priority)")
```
Renamed display name from "lowestPriorityNumber" to "should select node with highest priority" (clearer intent)
✅ Done!

AlfredoG87 commented on block-node/backfill/src/test/java/org/hiero/block/node/backfill/BackfillPersistenceAwaiterTest.java:
```diff
@@ -0,0 +1,412 @@
+        }
+
+        @Test
+        @DisplayName("should return true immediately if block not tracked")
```
Yes, the idea is:
true = "proceed, nothing is blocking you" and false = "timed out waiting".

If a block was never tracked, there's nothing to wait for - returning true correctly signals "no waiting needed, you can continue".

The alternative interpretation ("is this block definitely persisted?") would require checking the persistence layer, which isn't this class's responsibility.

The awaiter only tracks blocks it was explicitly asked to track, and once it receives persistenceNotification it forgets about it for saving memory.

AlfredoG87 commented on block-node/backfill/src/test/java/org/hiero/block/node/backfill/PriorityHealthBasedStrategyTest.java:
```diff
@@ -0,0 +1,282 @@
+
+            // Returns empty because only priority-1 node exists, and it's in backoff
+            assertTrue(result.isEmpty());
+        }
```
Returning empty triggers the retry mechanism in `backfillGap` of `BackfillRunner` class. The while loop calls `fetchNextChunk` which, on empty selection, replans availability by re-querying all nodes. If the backoff expired, the node becomes available and we retry. This keeps retrying until the preferred node recovers or we truly exhaust all options - rather than immediately falling back to lower priority nodes. This prevents from shifting the load to a lower priority node for a hiccup or transient error that quickly recovers on a high priority node, it also allows for returning to a high priority node faster since will Retry every max backoff periods if higher priority node is available.

Priority reflects prefference, a BN with higher priority could be cheaper or faster, so we want to make sure we always exhaust all the lower priority nodes, and that we come back to the highest priority nodes as soon as possible once they become available again.

AlfredoG87 commented on block-node/spi-plugins/src/main/java/org/hiero/block/node/spi/historicalblocks/LongRange.java:
```diff
@@ -122,6 +124,35 @@ public LongRange merge(LongRange other) {
+        merged.add(current);
+        return merged;
+    }
+
```
This has been discussed in the earlier thread (https://github.com/hiero-ledger/hiero-block-node/pull/2006#discussion_r2661554184) and I explained my reasoning (https://github.com/hiero-ledger/hiero-block-node/pull/2006#discussion_r2669973409). 

  I also addressed this in my https://github.com/hiero-ledger/hiero-block-node/pull/2006#issuecomment-3713067363:

  I've pondered this and decided that the benefits do not outweigh the effort and added complexity. The use case we need is simpler as opposed to the BlockRangeSet impls and prefer to keep it simple as a native List.

  The merge logic in LongRange is a simple static utility - no concurrency needed, single-threaded context, and avoids pulling in heavier dependencies. For now, to me, simplicity wins.

AlfredoG87 commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillFetcher.java:
```diff
@@ -0,0 +1,357 @@
+                    node.address(),
+                    node.port(),
+                    node.priority(),
+                    tuning != null ? "custom" : "defaults");
```
I like the PBJ generated toString, we lose some flexibility but we gain simplicity and readability of the codebase, seems like an acceptable trade-off. we get the needed information which is what matters in the end.

✅ Done!

jsync-swirlds commented on block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillFetcher.java:
```diff
@@ -0,0 +1,349 @@
+                        nodeConfig.address(),
+                        attempt,
+                        maxRetries,
+                        e.getMessage(),
```
I still advocate using `e` in preference to `e.getMessage()` in log messages.  Too many standard exceptions have no message (or an incomplete message) and rely on the type name (which is included in the `toString()` output) to make the "message" useable.

Non-blocking, however.

ata-nas commented on block-node/backfill/src/test/java/org/hiero/block/node/backfill/BackfillPluginTest.java:
```diff
@@ -39,8 +39,13 @@
+/**
+ * Integration tests for {@link BackfillPlugin}.
+ */
+@Timeout(value = 30, unit = TimeUnit.SECONDS)
```
Hmm, interesting. I see why you need the timeout now. But I do not think at all we can even remotely label these as integration tests. This test sever is in no way the real implementation, which is one of the conditions when having an integration tests. Curious, do we really need actual running servers, can't we have simple fixtures for these tests? Not asking for any changes, just thinking out loud.

AlfredoG87 commented on block-node/backfill/src/test/java/org/hiero/block/node/backfill/BackfillPluginTest.java:
```diff
@@ -1252,7 +1383,10 @@ public void handleVerification(VerificationNotification notification) {
+    /**
+     * Builder for creating backfill configuration maps for testing.
+     */
+    public static class BackfillConfigBuilder {
```
I used to have the alternative and it was a lot of repetitive code in every test, so I came up with this and has proven to be really useful and readable. I think we have different styles and neither is good or bad, but in this case is just a matter of personal preference.

Resolving this conversation. (feel free to bring it up to the parking lot discussion if you feel strongly about it, and you can reference this conversation for context)

AlfredoG87 commented on block-node/backfill/src/test/java/org/hiero/block/node/backfill/PriorityHealthBasedStrategyTest.java:
```diff
@@ -0,0 +1,282 @@
+            Optional<NodeSelection> result = strategy.select(0, 100, availability);
+
+            assertTrue(result.isPresent());
+            assertEquals(50, result.get().startBlock());
```
The process does eventually gets what it can, however it might take a couple passes, and we take a defensive approach, since we don't really know what it has in some cases when the source also has gaps.

Once the serverStatusDetails is implemented, we might be a bit more aggressive but with clearer information, and we might want to revisit backfill so we can improve it. until then, in all my testing, this is working reliable and quite well.

Also resolving this conversation.
